### PR TITLE
Offload diagnostics

### LINKED ISSUE #34: Offload Diagnostics
<!--

Thanks for opening an issue! To help the Numba team handle your information
efficiently, please first ensure that there is no other issue present that
already describes the issue you have
(search at https://github.com/numba/numba/issues?&q=is%3Aissue).

-->

## Feature request

<!--

Please include details of the feature you would like to see, why you would
like to see it/the use case.

-->

Currently in Numba we get information about which code region was identified as parallel through the use of NUMBA_PARALLEL_DIAGNOSTICS.

e.g.

```
import numba                                                                                                                                                                                                                                 import numpy as np                                                                                                                                                                                                                           import math                                                                                                                                                                                                                                  import argparse                                                                                                                                                                                                                              import time                                                                                                                                                                                                                                  import dpctl

@numba.vectorize(nopython=True)
def cndf2(inp):
    out = 0.5 + 0.5 * math.erf((math.sqrt(2.0)/2.0) * inp)
    return out

@numba.njit(parallel=True, fastmath=True)
def blackscholes(sptprice, strike, rate, volatility, timev):
    logterm = np.log(sptprice / strike)
    powterm = 0.5 * volatility * volatility
    den = volatility * np.sqrt(timev)
    d1 = (((rate + powterm) * timev) + logterm) / den
    d2 = d1 - den
    NofXd1 = cndf2(d1)
    NofXd2 = cndf2(d2)
    futureValue = strike * np.exp(- rate * timev)
    c1 = futureValue * NofXd2
    call = sptprice * NofXd1 - c1
    put  = call - futureValue + sptprice
    return put


def run(iterations):
    sptprice   = np.full((iterations,), 42.0)
    initStrike = 40 + (np.arange(iterations) + 1.0) / iterations
    rate       = np.full((iterations,), 0.5)
    volatility = np.full((iterations,), 0.2)
    timev      = np.full((iterations,), 0.5)

    t1 = time.time()
    put = blackscholes(sptprice, initStrike, rate, volatility, timev)
    t = time.time()-t1
    print("checksum: ", sum(put))
    print("SELFTIMED ", t)

def main():
    parser = argparse.ArgumentParser(description='Black-Scholes')
    parser.add_argument('--options', dest='options', type=int, default=10000000)
    args = parser.parse_args()
    options = args.options

    run(10)
    print("options = ", options)
    run(options)

if __name__ == '__main__':
    main()
```

When run with the following command

` NUMBA_PARALLEL_DIAGNOSTICS=1 python example.py`

Yields the following output:

```
================================================================================
 Parallel Accelerator Optimizing:  Function blackscholes,
numba/dppl/examples/blacksholes_njit.py (18)
================================================================================


Parallel loop listing for  Function blackscholes, numba/dppl/examples/blacksholes_njit.py (18)
----------------------------------------------------------------|loop #ID
@numba.njit(parallel=True, fastmath=True)                       |
def blackscholes(sptprice, strike, rate, volatility, timev):    |
    logterm = np.log(sptprice / strike)-------------------------| #0
    powterm = 0.5 * volatility * volatility---------------------| IntelPython/numba#1
    den = volatility * np.sqrt(timev)---------------------------| IntelPython/numba#2
    d1 = (((rate + powterm) * timev) + logterm) / den-----------| IntelPython/numba#3
    d2 = d1 - den-----------------------------------------------| IntelPython/numba#4
    NofXd1 = cndf2(d1)------------------------------------------| IntelPython/numba#5
    NofXd2 = cndf2(d2)------------------------------------------| IntelPython/numba#6
    futureValue = strike * np.exp(- rate * timev)---------------| IntelPython/numba#7
    c1 = futureValue * NofXd2-----------------------------------| IntelPython/numba#8
    call = sptprice * NofXd1 - c1-------------------------------| IntelPython/numba#9
    put  = call - futureValue + sptprice------------------------| IntelPython/numba#10
    return put                                                  |
------------------------------ After Optimisation ------------------------------
Parallel region 0:
+--0 (parallel, fused with loop(s): 1, 2, 3, 4, 5, 6, 7, 8, 9, 10)



Parallel region 0 (loop #0) had 10 loop(s) fused.
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
```

We want to generate a similar report for auto-offload too using a new environment flag `DPPY_OFFLOAD_DIAGNOSTICS`. We should also support the "verbosity levels" like the `NUMBA_PARALLEL_DIAGNOSTICS=1`, *i.e.*, `DPPY_OFFLOAD_DIAGNOSTICS=1` or `DPPY_OFFLOAD_DIAGNOSTICS=2` *etc*. To begin with, we will support only one level.

## Expected output when all regions could be offloaded:

```
================================================================================
 Offload Diagnostics:  Function blackscholes,
numba/dppl/examples/blacksholes_njit.py (18)
================================================================================


Parallel loop listing for  Function blackscholes, numba/dppl/examples/blacksholes_njit.py (18)
----------------------------------------------------------------|loop #ID
@numba.njit(parallel=True, fastmath=True)                       |
def blackscholes(sptprice, strike, rate, volatility, timev):    |
    logterm = np.log(sptprice / strike)-------------------------| #0
    powterm = 0.5 * volatility * volatility---------------------| IntelPython/numba#1
                                                                |
    print("First Break")                                        |
                                                                |
    den = volatility * np.sqrt(timev)---------------------------| IntelPython/numba#2
    d1 = (((rate + powterm) * timev) + logterm) / den-----------| IntelPython/numba#3
                                                                |
    print("Second Break")                                       |
                                                                |
    d2 = d1 - den-----------------------------------------------| IntelPython/numba#4
    NofXd1 = cndf2(d1)------------------------------------------| IntelPython/numba#5
    NofXd2 = cndf2(d2)------------------------------------------| IntelPython/numba#6
    futureValue = strike * np.exp(- rate * timev)---------------| IntelPython/numba#7
    c1 = futureValue * NofXd2-----------------------------------| IntelPython/numba#8
    call = sptprice * NofXd1 - c1-------------------------------| IntelPython/numba#9
    put  = call - futureValue + sptprice------------------------| IntelPython/numba#10
                                                                |
    return put                                                  |
------------------------------ After Optimisation ------------------------------
Parallel region 0:
+--0 (parallel, fused with loop(s): 1)


Parallel region 1:
+--2 (parallel, fused with loop(s): 3)


Parallel region 2:
+--4 (parallel, fused with loop(s): 5, 6, 7, 8, 9, 10)



Parallel region 0 (loop #0) had 1 loop(s) fused.

Parallel region 1 (loop IntelPython/numba#2) had 1 loop(s) fused.

Parallel region 2 (loop IntelPython/numba#4) had 6 loop(s) fused.
------------------------------ Auto-offloading ------------------------------
Parallel region 0 offloaded to:
+--0 (parallel, fused with loop(s): 1)

Device - "actual device name from dpctl"
Global Size -
Local Size - Auto

Parallel region 1 offloaded to:
+--2 (parallel, fused with loop(s): 3)

Device - "actual device name from dpctl"
Global Size -
Local Size - Auto

Parallel region 2 offloaded to:
+--4 (parallel, fused with loop(s): 5, 6, 7, 8, 9, 10)

Device - "actual device name from dpctl"
Global Size -
Local Size - Auto
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
```

## Expected output when one kernel could not be offloaded


```
================================================================================
 Offload Diagnostics:  Function blackscholes,
numba/dppl/examples/blacksholes_njit.py (18)
================================================================================


Parallel loop listing for  Function blackscholes, numba/dppl/examples/blacksholes_njit.py (18)
----------------------------------------------------------------|loop #ID
@numba.njit(parallel=True, fastmath=True)                       |
def blackscholes(sptprice, strike, rate, volatility, timev):    |
    logterm = np.log(sptprice / strike)-------------------------| #0
    powterm = 0.5 * volatility * volatility---------------------| IntelPython/numba#1
                                                                |
    print("First Break")                                        |
                                                                |
    den = volatility * np.sqrt(timev)---------------------------| IntelPython/numba#2
    d1 = (((rate + powterm) * timev) + logterm) / den-----------| IntelPython/numba#3
                                                                |
    print("Second Break")                                       |
                                                                |
    d2 = d1 - den-----------------------------------------------| IntelPython/numba#4
    NofXd1 = cndf2(d1)------------------------------------------| IntelPython/numba#5
    NofXd2 = cndf2(d2)------------------------------------------| IntelPython/numba#6
    futureValue = strike * np.exp(- rate * timev)---------------| IntelPython/numba#7
    c1 = futureValue * NofXd2-----------------------------------| IntelPython/numba#8
    call = sptprice * NofXd1 - c1-------------------------------| IntelPython/numba#9
    put  = call - futureValue + sptprice------------------------| IntelPython/numba#10
                                                                |
    return put                                                  |
------------------------------ After Optimisation ------------------------------
Parallel region 0:
+--0 (parallel, fused with loop(s): 1)


Parallel region 1:
+--2 (parallel, fused with loop(s): 3)


Parallel region 2:
+--4 (parallel, fused with loop(s): 5, 6, 7, 8, 9, 10)



Parallel region 0 (loop #0) had 1 loop(s) fused.

Parallel region 1 (loop IntelPython/numba#2) had 1 loop(s) fused.

Parallel region 2 (loop IntelPython/numba#4) had 6 loop(s) fused.
------------------------------ Auto-offloading ------------------------------
Parallel region 0 offloaded to:
+--0 (parallel, fused with loop(s): 1)

Device - "actual device name from dpctl"
Global Size -
Local Size - Auto

Parallel region 1 offloaded to:
+--2 (parallel, fused with loop(s): 3)

Device - Failed to offload to a device! {PRINT ERROR_MSG WHY IT FAILED}
Global Size -
Local Size - Auto

Parallel region 2 offloaded to:
+--4 (parallel, fused with loop(s): 5, 6, 7, 8, 9, 10)

Device - Failed to offload to a device since Parallel region 1 failed.
Global Size -
Local Size - Auto
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
```

### PR DESCRIPTION
**Pull request by 1e-to**
Closes issue https://github.com/IntelPython/numba/issues/113
Closes #34

### CODE REVIEW

**Comment by PokhodenkoSA** [comment]
@reazulhoque please, review it.

File: numba_dppy/dppy_parfor_diagnostics.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        if self.extra_info:
+            parfors_simple = self.get_parfors_simple(False)
+            all_lines = self.get_all_lines(parfors_simple)
+            print(' Auto-offloading '.center(_termwidth,'-'))
+            self.print_auto_offloading(all_lines)
+            if 'kernel' in self.extra_info.keys():
+                print_wrapped("Device - '%s'" % self.extra_info['kernel'])
+            print(_termwidth * '-')
+
+    def print_auto_offloading(self, lines):
```
Comment (reazulhoque) [inline numba_dppy/dppy_parfor_diagnostics.py:21]: Are we overriding these methods? Is it to have more control on what we are printing?

File: numba_dppy/tests/test_offload_diagnostics.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+def driver(a, b, c, global_size):
+    data_parallel_sum[global_size, dppy.DEFAULT_LOCAL_SIZE](a, b, c)
+
+
+@unittest.skipUnless(dpctl.has_gpu_queues(), "test only on GPU system")
+class TestOffloadDiagnostics(DPPYTestCase):
+    def test_parfor(self):
+        with dpctl.device_context("opencl:gpu"):
+            dppy_config.OFFLOAD_DIAGNOSTICS = 1
+            jitted = njit(parallel={"offload": True})(prange_example)
```
Comment (reazulhoque) [inline numba_dppy/tests/test_offload_diagnostics.py:37]: We should not need the offload option anymore

File: numba_dppy/dppy_parfor_diagnostics.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
+from numba.parfors.parfor import ParforDiagnostics, _termwidth, print_wrapped
+
+
+class ExtendedParforDiagnostics(ParforDiagnostics):
+    def __init__(self):
+        ParforDiagnostics.__init__(self)
+        self.extra_info = {}
+
+    def dump(self, level=1):
+        ParforDiagnostics.dump(self, level=1)
```
Comment (reazulhoque) [inline numba_dppy/dppy_parfor_diagnostics.py:10]: Why are we hardcoding level to 1 here?

File: numba_dppy/compiler.py [hunk 4]
```diff
@@ -58,6 +59,8 @@ class DPPYCompiler(CompilerBase):
     def define_pipelines(self):
         # this maintains the objmode fallback behaviour
         pms = []
+        self.state.parfor_diagnostics = ExtendedParforDiagnostics()
+        self.state.metadata['parfor_diagnostics'] = self.state.parfor_diagnostics
```
Comment (reazulhoque) [inline numba_dppy/compiler.py:63]: Is `'parfor_diagnostics'`  attached to the `state.metadata` or `lowerer.metadata` or both? If both can you please explain why?

**Review by reazulhoque — COMMENTED** [review]
This work will provide users with priliminary information like which functions were offloaded to th GPU with the specific device name in case there are multiple devices. We need more work to show additional information like global size and local size etc.

@diptorupd Is showing the device name along with which kernel was actually offloaded enough for our next milestone? I personally think it is. Please let us know.

After the comments are addressed this should be good to merge if CI passes.

File: numba_dppy/dppy_lowerer.py [hunk 5]
```diff
@@ -1165,6 +1167,9 @@ def lower(self):
         try:
             lowering.lower_extensions[parfor.Parfor].append(lower_parfor_rollback)
             self.gpu_lower.lower()
+            # if lower dont crash, and parfor_diagnostics is empthy then it is kernel
```
Comment (PokhodenkoSA) [inline numba_dppy/dppy_lowerer.py:1170]:
```suggestion
            # if lower dont crash, and parfor_diagnostics is empty then it is kernel
```

File: numba_dppy/dppy_lowerer.py [hunk 6]
```diff
@@ -1165,6 +1167,9 @@ def lower(self):
...
+            if not self.gpu_lower.metadata['parfor_diagnostics'].extra_info:
+                self.gpu_lower.metadata['parfor_diagnostics'].extra_info["kernel"] = str(dpctl.get_current_queue().get_sycl_device().get_device_name())
```
Comment (PokhodenkoSA) [inline numba_dppy/dppy_lowerer.py:1172]: Very long string. black is coming :)

File: numba_dppy/dppy_lowerer.py (hunk 5 above)
Comment (PokhodenkoSA) [reply numba_dppy/dppy_lowerer.py:1170]: @reazulhoque you can propose code change via 
image

File: numba_dppy/dppy_parfor_diagnostics.py (hunk 3 above)
Comment (1e-to) [reply numba_dppy/dppy_parfor_diagnostics.py:10]: In numba's dump function there is should be defined level
If we don't do it, level=0 and "Report level unknown, should be one of 1, 2, 3, 4"

File: numba_dppy/compiler.py (hunk 4 above)
Comment (1e-to) [reply numba_dppy/compiler.py:63]: Parfor diagnostics is attached to the state.metadata, and then in SpirvFriendlyLowering it is transferred to DPPYLower
That is, the same diagnostics is used in the lowerer.

Comment (reazulhoque) [reply numba_dppy/compiler.py:63]: Makes sense

File: numba_dppy/dppy_parfor_diagnostics.py (hunk 3 above)
Comment (reazulhoque) [reply numba_dppy/dppy_parfor_diagnostics.py:10]: level is set to 1 by default in line 9, can we not pass that to ParforDiagnostics?

Comment (1e-to) [reply numba_dppy/dppy_parfor_diagnostics.py:10]: When we specify the level value in the environment variable, then it is transmitted, and by default (if nothing is specified) level 1 works

Comment (AlexanderKalistratov) [reply numba_dppy/dppy_parfor_diagnostics.py:10]: @1e-to I think comment is about line 10
```python
ParforDiagnostics.dump(self, level=1)
```
Shouldn't it be:
```python
ParforDiagnostics.dump(self, level=level)
```
?

Comment (1e-to) [reply numba_dppy/dppy_parfor_diagnostics.py]: Fixed

Comment (AlexanderKalistratov) [reply numba_dppy/dppy_parfor_diagnostics.py]: Also isn't it:

```python
super().dump(level)
```

?

**Pull request merged**